spacescout_labstats/test/test_cte_techloan.py

```diff
@@ -0,0 +1,275 @@
+        self.assertEqual(
+            'i_reservation_required' in test_items[1]["extended_info"], False)
+
+    def test_get_techloan_data_by_id(self):
+        """
+        Tests that the returned techloan_data has a equipment_location_id
+        that matches the given id. Also tests that an empty list is
+        returned when not found
+        """
+        # Test with a list of data that have the matching techloan_id
+        valid_data = self.load_json_file('cte_techloan_type_data.json')
+        returned = cte_techloan.get_techloan_data_by_id(valid_data, 1)
+        # Should return a list with data that has the correct id
+        self.assertEqual(valid_data, returned)
+
+        # Test with a list with data thats valid but doesnt match the id
+        valid_data[0]["equipment_location_id"] = 2
+        returned = cte_techloan.get_techloan_data_by_id(valid_data, 1)
+        self.assertEqual(returned, [])
+
+    def test_get_space_item_by_type_id(self):
+        """
+        Tests that the correct space with a cte_type_id that matches the given
+        type_id. Also tests that 'None' is returned when not found
+        """
+        # Test with a list of spaces that have the matching cte_type_id
+        valid_spaces = self.load_json_file('cte_techloan.json')
+        items = valid_spaces[0]["items"]
+        returned = cte_techloan.get_space_item_by_type_id(1, items)
+        # Should return a single space with the correct id
+        self.assertEqual(items[0], returned)
+
+        # Test with a list of spaces that are valid but dont have a matching id
+        items[0]["extended_info"]["cte_type_id"] = 2
+        returned = cte_techloan.get_space_item_by_type_id(1, items)
+        self.assertIs(returned, None)
+
+        # Test with a list of spaces that do not have a cte_type_id
+        invalid_items = [{}]
+        invalid_items[0]["extended_info"] = {}
+        returned = cte_techloan.get_space_item_by_type_id(1, invalid_items)
+        self.assertIs(returned, None)
+
+    def test_validate_techloan_data(self):
+        """
+        Tests that techloan data is validated appropriately and that
+        exceptions are raised with invalid techloan data versions
+        """
+        valid_type_data = self.load_json_file('cte_techloan_type_data.json')
+        # Test validation with data that is of dict type instead of list
+        invalid_type_data = {}
+        with self.assertRaises(Exception):
+            cte_techloan.validate_techloan_data(invalid_type_data)
+
+        invalid_entry_data = [{}]
+        # Test validation with data of correct type containing an invalid entry
+        # This should remove the single invalid entry in the invalid entry data
+        cte_techloan.validate_techloan_data(invalid_entry_data)
+        self.assertEqual(invalid_entry_data, [])
+
+        # Test with data that is of correct type with a valid entry
+        cte_techloan.validate_techloan_data(valid_type_data)
```

Comment:
Assert that nothing was logged/removed after this step, as invalid data should not be modified in any way